pkg/slayers/path/scion/raw.go

```diff
@@ -397,54 +398,68 @@ func (s *Raw) GetCurrentInfoField( /*@ ghost ubuf []byte @*/ ) (res path.InfoFie
 // @ 	validPktMetaHdr(ubuf) && s.EqAbsHeader(ubuf)
 // @ ensures  r == nil && idx == int(old(s.GetCurrINF(ubuf))) ==>
 // @ 	let oldPkt := old(s.absPkt(ubuf)) in
 // @ 	let newPkt := AbsSetInfoField(oldPkt, info.ToIntermediateAbsInfoField()) in
 // @ 	s.absPkt(ubuf) == newPkt
 // @ decreases
+// @ #backend[exhaleMode(1)]
 func (s *Raw) SetInfoField(info path.InfoField, idx int /*@, ghost ubuf []byte @*/) (r error) {
 	//@ share info
-	//@ ghost oldCurrINF := int(old(s.GetCurrINF(ubuf)))
+	//@ reveal validPktMetaHdr(ubuf)
 	//@ unfold acc(s.Mem(ubuf), R50)
 	//@ unfold acc(s.Base.Mem(), R50)
+	//@ currInfIdx := int(s.PathMeta.CurrINF)
+	//@ currHfIdx := int(s.PathMeta.CurrHF)
+	//@ seg1Len := int(s.PathMeta.SegLen[0])
+	//@ seg2Len := int(s.PathMeta.SegLen[1])
+	//@ seg3Len := int(s.PathMeta.SegLen[2])
+	//@ segLen := LengthOfCurrSeg(currHfIdx, seg1Len, seg2Len, seg3Len)
+	//@ prevSegLen := LengthOfPrevSeg(currHfIdx, seg1Len, seg2Len, seg3Len)
+	//@ offset := HopFieldOffset(s.Base.NumINF, prevSegLen, MetaLen)
+	//@ hopfieldOffset := MetaLen + s.NumINF*path.InfoLen
 	if idx >= s.NumINF {
 		err := serrors.New("InfoField index out of bounds", "max", s.NumINF-1, "actual", idx)
 		//@ fold acc(s.Base.Mem(), R50)
 		//@ fold acc(s.Mem(ubuf), R50)
 		return err
 	}
 	infOffset := MetaLen + idx*path.InfoLen
-	//@ assert idx == oldCurrINF ==> reveal validPktMetaHdr(ubuf)
-	//@ assert idx == oldCurrINF ==> s.EqAbsHeader(ubuf)
-
-	//@ sl.SplitRange_Bytes(ubuf, 0, len(s.Raw), HalfPerm)
-	//@ ValidPktMetaHdrSublice(ubuf, len(s.Raw))
-	//@ sl.SplitRange_Bytes(ubuf, 0, len(s.Raw), HalfPerm)
-	//@ assert idx == oldCurrINF ==> RawBytesToBase(ubuf[:len(s.Raw)]).ValidCurrIdxsSpec()
 
-	//@ assert sl.AbsSlice_Bytes(s.Raw, 0, len(s.Raw))
-	//@ sl.SplitRange_Bytes(s.Raw, infOffset, infOffset+path.InfoLen, HalfPerm)
-	//@ assert acc(sl.AbsSlice_Bytes(s.Raw, 0, infOffset), HalfPerm)
-	//@ sl.Reslice_Bytes(s.Raw, 0, infOffset, HalfPerm/2)
-	//@ ValidPktMetaHdrSublice(s.Raw, infOffset)
-	//@ sl.SplitRange_Bytes(s.Raw, infOffset, infOffset+path.InfoLen, HalfPerm)
-	//@ assert idx == oldCurrINF ==> RawBytesToBase(s.Raw[:infOffset]).ValidCurrIdxsSpec()
+	//@ SliceBytesbyInfoFields(ubuf, MetaLen, s.NumINF, HalfPerm)
+	//@ SliceBytesbySegments(ubuf[hopfieldOffset:], 0, seg1Len, seg2Len, seg3Len, HalfPerm)
 
+	//@ ValidPktMetaHdrSublice(ubuf, MetaLen)
+	//@ oldInfo := path.BytesToIntermediateAbsInfoField(ubuf[infOffset : infOffset+path.InfoLen], 0, 0, path.InfoLen)
+	//@ newInfo := info.ToIntermediateAbsInfoField()
+	//@ hfIdxSeg := currHfIdx-prevSegLen
+	//@ hopfields := ubuf[offset:offset + segLen*path.HopLen]
+	//@ ghost if(idx == currInfIdx) {
+	//@ 	CurrSegEqLemma(ubuf, offset, currInfIdx, hfIdxSeg, segLen)
+	//@ 	LeftSegEqLemma(ubuf, currInfIdx+1, seg1Len, seg2Len, seg3Len)
+	//@  	MidSegEqLemma(ubuf, currInfIdx+2, seg1Len, seg2Len, seg3Len)
+	//@ 	RightSegEqLemma(ubuf, currInfIdx-1, seg1Len, seg2Len, seg3Len)
+	//@ }
+	//@ reveal s.absPkt(ubuf)
```

Comment:
here, do we actually need to review the full function, or is there any specific property that we want to obtain from this? Maybe introducing a separate lemma for the property we need helps with perf, at the cost of more annotation overhead

Reply:
Unfortunately, we require the full function. Additionally, I don't think `absPkt` would increase perf that much because it  uses opaque quite heavily.